feat(listAllDependencies): list all dependencies in output

CedricGatay opened the pull request:
Allow listing all dependencies in output:
 * CLI switch added to allow this (`-u` / `--include-up-to-date`)
 * Markdown output updated
 * JSON output contains `upToDatePackages` root object
 * XCode output untouched, as it makes no sense

CedricGatay committed:
feat(listAllDependencies): list all dependencies in output

Allow listing all dependencies in output:
 * CLI switch added to allow this (`-u` / `--include-up-to-date`)
 * Markdown output updated
 * JSON output contains `currentPackages` root object
 * XCode output untouched, as it makes no sense

kiliankoe commented on Sources/Outdated/PackageCollection.swift:
```diff
@@ -27,24 +28,40 @@ extension PackageCollection {
+            .dropFirst()
+            .dropLast(1)
+            .joined(separator: "\n")
+    }
```
This always returns a rendered table, even if the list is empty, which feels a bit awkward for ignored packages. Do you think for the case of an empty table it should say there's no content or skip that section entirely?

```
## Outdated packages
| Package       | Current | Latest                        | URL                                           |
|---------------|---------|-------------------------------|-----------------------------------------------|
| rainbow       | 3.2.0   | 4.2.0                         | https://github.com/onevcat/Rainbow.git        |
| swift-syntax  | 601.0.1 | 603.0.0-prerelease-2025-08-11 | https://github.com/swiftlang/swift-syntax.git |
| swift-testing | 6.1.1   | 6.1.2                         | https://github.com/apple/swift-testing.git    |
## Up to date packages
| Package               | Current | URL                                                |
|-----------------------|---------|----------------------------------------------------|
| files                 | 4.3.0   | https://github.com/johnsundell/Files.git           |
| shellout              | 2.3.0   | https://github.com/johnsundell/shellout.git        |
| swift-argument-parser | 1.6.1   | https://github.com/apple/swift-argument-parser.git |
| swift-log             | 1.6.4   | https://github.com/apple/swift-log.git             |
| swiftytexttable       | 0.9.0   | https://github.com/scottrhoyt/swiftytexttable.git  |
| version               | 2.2.0   | https://github.com/mxcl/Version.git                |
## Ignored packages
| Package | Current | URL |
|---------|---------|-----|
```

CedricGatay replied:
I think I need to skip the title and table display, you're absolutely right about this. I'll update my PR (as well as not displaying title for outdated packages if there is none.

kiliankoe reviewed (commented):
Thanks for the PR! Seems like a nice addition, but I'm curious what your use case for this flag is? 

I left an inline comment regarding markdown table output. The rest looks good to me 👍

CedricGatay commented:
The use case is for reporting the list of currently used dependencies for every releases of my project, this is required by my global security team

CedricGatay committed:
feat(listAllDependencies): skip empty section

If a outdated or ignored does not contain any package, skip its output entirely (including title)

kiliankoe reviewed (commented):
Maybe it would make sense to move the section title into `render` as an argument and then have that function decide to either return a table or not? That way it would also handle a potentially empty list of outdated or up-to-date packages correctly (as in not showing them).

kiliankoe commented:
That use case makes a lot of sense, thanks for the insight! It really helps to understand the why behind a feature!

CedricGatay commented:
> Maybe it would make sense to move the section title into `render` as an argument and then have that function decide to either return a table or not? That way it would also handle a potentially empty list of outdated or up-to-date packages correctly (as in not showing them).

I am adding this, this is a good point

CedricGatay committed:
feat(listAllDependencies): render responsible for title

Render is now responsible for title output if there is one defined
Render now prints to the console instead of returning the formatted value
to print later on

CedricGatay committed:
feat(listAllDependencies): make *Package struct Comparable

Implement Comparable on *Package structs to be able to use a single `sorted()` call instead of duplicating the sort closure.
Sorted ignoredPackages as well

CedricGatay commented:
Fixed with the render method being reponsible for title / table printing to stdio and implement `Comparable` on Packages struct to make sorting cleaner to read.

kiliankoe commented:
Fantastic, thank you for the contribution!

kiliankoe commented:
It's included in 0.10.0 and should be available via homebrew soon 👍
